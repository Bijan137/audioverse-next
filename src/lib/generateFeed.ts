import { Feed } from 'feed';

import { GenerateFeedFragment } from '@lib/generated/graphql';

import getIntl from './getIntl';
import getLanguageByBaseUrl from './getLanguageByBaseUrl';

export function generateFeed(
	feedTitle: string,
	recordings: GenerateFeedFragment[],
	languageRoute: string
): string {
	const intl = getIntl(languageRoute);
	const language = getLanguageByBaseUrl(languageRoute);

	// TODO: Switch to automatic ID generation:
	// https://formatjs.io/docs/getting-started/message-extraction/#automatic-id-generation
	const title = intl.formatMessage(
		{
			id: 'feed-title',
			defaultMessage: '{identifier} | AudioVerse {lang}',
			description: 'Generic feed title',
		},
		{
			identifier: feedTitle,
			lang: language?.display_name,
		}
	);

	// https://github.com/avorg/wp-avorg-plugin/blob/master/view/page-feed.twig
	// TODO: copyright
	// TODO: itunes:subtitle
	// TODO: id?
	// TODO: itunes:explicit
	// TODO: link
	// TODO: atom:link
	// TODO: itunes:author
	// TODO: image
	// TODO: categories
	const feed = new Feed({
		id: '',
		title,
		copyright: '',
	});

	recordings.map((r) => {
		const { audioFiles = [], feedVideoFiles = [] } = r;
		const files = [...audioFiles, ...feedVideoFiles];
		const file = files.length && files[0];

		if (!file) return;

		const url = file.url;
		const length = parseInt(file.filesize);

		if (!url) return;

		// https://github.com/avorg/wp-avorg-plugin/blob/master/view/page-feed.twig
		// TODO: Add image
		// TODO: Add itunes:subtitle
		// TODO: Add itunes:author
		// TODO: Add pubDate (?)
		// TODO: Add itunes:duration
		feed.addItem({
			title: r.title,
			description: r.description || undefined,
			link: r.canonicalUrl,
			date: new Date(r.recordingDate || ''),
			enclosure: { url, length },
		});
	});

	return feed.rss2();
}
