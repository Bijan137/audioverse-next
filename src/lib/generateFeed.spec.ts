import { advanceTo } from 'jest-date-mock';

import { generateFeed } from './generateFeed';

describe('generateFeed', () => {
	beforeAll(() => {
		advanceTo(new Date(2021, 8, 13, 12, 34, 56));
	});
	it('makes feed', () => {
		const result = generateFeed('', [], 'en');
		expect(result).toMatchSnapshot();
	});

	it('includes feed title', () => {
		const result = generateFeed('the_title', [], 'en');
		expect(result).toMatchSnapshot();
	});

	it('includes adds item', () => {
		const result = generateFeed(
			'the_title',
			[
				{
					title: 'recording_title',
					description: 'recording_description',
					recordingDate: '2007-03-05T12:00:00.000Z',
					canonicalUrl: 'the_url',
					audioFiles: [
						{
							url: 'https://www.example.com/file_url',
							filesize: '3',
						},
					],
					feedVideoFiles: [],
				},
			],
			'en'
		);
		expect(result).toMatchSnapshot();
	});
});
