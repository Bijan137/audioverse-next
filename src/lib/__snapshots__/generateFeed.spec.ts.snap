// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFeed includes adds item 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>the_title | AudioVerse English</title>
        <link>undefined</link>
        <description>undefined</description>
        <lastBuildDate>Mon, 13 Sep 2021 12:34:56 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/jpmonette/feed</generator>
        <item>
            <title><![CDATA[recording_title]]></title>
            <link>the_url</link>
            <guid>the_url</guid>
            <pubDate>Mon, 05 Mar 2007 12:00:00 GMT</pubDate>
            <description><![CDATA[recording_description]]></description>
            <enclosure length=\\"3\\" type=\\"image//file_url\\" url=\\"https://www.example.com/file_url\\"/>
        </item>
    </channel>
</rss>"
`;

exports[`generateFeed includes feed title 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>the_title | AudioVerse English</title>
        <link>undefined</link>
        <description>undefined</description>
        <lastBuildDate>Mon, 13 Sep 2021 12:34:56 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/jpmonette/feed</generator>
    </channel>
</rss>"
`;

exports[`generateFeed makes feed 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title> | AudioVerse English</title>
        <link>undefined</link>
        <description>undefined</description>
        <lastBuildDate>Mon, 13 Sep 2021 12:34:56 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/jpmonette/feed</generator>
    </channel>
</rss>"
`;
